src/fpm/manifest/profiles.f90

```diff
@@ -0,0 +1,879 @@
+
+        !> List of profiles
+        type(profile_config_t), allocatable, intent(inout), optional :: profiles(:)
+
+        !> Index in the list of profiles
+        integer, intent(inout), optional :: profindex
+        
+        type(toml_key), allocatable :: key_list(:)
+        character(len=:), allocatable :: os_name, l_os_name
+        type(toml_table), pointer :: os_node
+        character(len=:), allocatable :: flags
+        integer :: ios, stat, os_type
+        logical :: is_valid, key_val_added, is_key_val
+
+        if (size(os_list)<1) return
+        key_val_added = .false.
+        do ios = 1, size(os_list)
+          os_name = os_list(ios)%key
+          call validate_os_name(os_name, is_valid)
+          if (is_valid) then
+            call get_value(table, os_name, os_node, stat=stat)
+            if (stat /= toml_stat%success) then
+              call syntax_error(error, "os "//os_name//" has to be a table")
+              return
+            end if
+            call os_node%get_keys(key_list)
+            if (present(profiles_size)) then
+              profiles_size = profiles_size + 1
+              call validate_profile_table(profile_name, compiler_name, key_list, os_node, error, .true.)
+            else
+              if (.not.(present(profiles).and.present(profindex))) then
+                call fatal_error(error, "Both profiles and profindex have to be present")
+                return
+              end if
+              call match_os_type(os_name, os_type)
+              call get_flags(profile_name, compiler_name, os_type, key_list, os_node, profiles, profindex, .true.)
+            end if
+          else
+            ! Not lowercase OS name
+            l_os_name = lower(os_name)
+            call validate_os_name(l_os_name, is_valid)
+            if (is_valid) then
+              call fatal_error(error,'*traverse_oss*:Error: Invalid OS name.')
+            end if
+            if (allocated(error)) return
+
+            ! Missing OS name
+            is_key_val = .false.
+            os_name = os_list(ios)%key
+            call get_value(table, os_name, os_node, stat=stat)
+            if (stat /= toml_stat%success) then
+              is_key_val = .true.
+            end if
+            os_node=>table
+            if (present(profiles_size)) then
+              if (is_key_val.and..not.key_val_added) then
+                key_val_added = .true.
+                is_key_val = .false.
+                profiles_size = profiles_size + 1
+              else if (.not.is_key_val) then
+                profiles_size = profiles_size + 1
+              end if
+              call validate_profile_table(profile_name, compiler_name, os_list, os_node, error, .false.)
+            else
+              if (.not.(present(profiles).and.present(profindex))) then
+                call fatal_error(error, "Both profiles and profindex have to be present")
+                return
+              end if
```

Comment:
Is this an internal error or can it be triggered by the user? In case this is user visible (let's assume it is), how helpful would that message be for fixing the manifest?

Reply:
It is an internal error

Reply:
So if the routine cannot be called with the optional arguments absent, can't we just always require them to be present?

Reply:
Oh, the thing is, the function operates in two modes, one is for just getting to know how many profiles there are if I am correct and the other actually writes to the profiles array.
That's where the line 403 and 431 comes in with `if (present(profiles_size))`.

Reply:
I see. There are 3 optional arguments. The procedure should be called with either just `profiles_size` present, or both `profiles` and `profindex` present. Would it make sense to split this into 2 procedures where neither has optional arguments?

Reply:
@everythingfunctional @awvwgk In the latest commit, I have split the function into two. One is `traverse_oss_for_size` the other is just `traverse_oss`. I tried to remove all unnecessary variables. Let me know what you think.